### PR TITLE
Allow PD gains to be set through datastore call

### PR DESCRIPTION
**Pull request by rohanpsingh**
~[this is currently a draft which allows pd gains to be set only when the OutPorts are available (i.e. when using ModifiedServo)]~

This PR will create datastore functions to modify the servo PD gains, either directly on the iob if `REAL_ROBOT = ON`  or through RTC ports if `REAL_ROBOT = OFF`.`REAL_ROBOT` is a cmake option.

The datastores can be called from an mc-rtc controller as follows:
- Example for reading pd gains:
```
  std::vector<double> kp;
  std::vector<double> kd;
  if(ctl.datastore().has(ctl.robot().name() + "::GetPDGains"))
  {
    bool get_servo_success = ctl.datastore().call<bool, std::vector<double> &, std::vector<double> &>(
        ctl.robot().name() + "::GetPDGains", kp, kd);
  }
```

- Example for writing pd gains:
```
  if(ctl.datastore().has(ctl.robot().name() + "::SetPDGains"))
  {
    bool set_servo_success = ctl.datastore().call<bool, const std::vector<double> &, const std::vector<double> &>(
        ctl.robot().name() + "::SetPDGains", kp, kd);
  }
  ```

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -34,3 +34,5 @@ add_subdirectory(idl)
 
 add_subdirectory(projects)
 add_subdirectory(src)
+
+option(REAL_ROBOT "If this is ON, mc_openrtm will attempt to change servo PD gains through IOB." OFF)
```
Comment (gergondet) [inline CMakeLists.txt:38]: I'd prefer this to be a run-time switch. We can easily detect if mc_openrtm is being run on a real robot (`RobotHardware` component is available) or in simulation

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -24,8 +24,12 @@ add_library(MCControl SHARED ${MCControl_SRC} ${MCControl_HDR})
 add_dependencies(MCControl generate_mc_idl_cpp)
 set_target_properties(MCControl PROPERTIES PREFIX "")
 pkg_config_use_dependency(MCControl hrpsys-base)
-target_link_libraries(MCControl PUBLIC mc_rtc::mc_control)
-target_include_directories(MCControl PRIVATE "${PROJECT_BINARY_DIR}/idl")
+target_link_libraries(MCControl PUBLIC mc_rtc::mc_control "${HRPSYS_BASE_PREFIX}/lib/libhrpIo.so")
+target_include_directories(MCControl PRIVATE "${PROJECT_BINARY_DIR}/idl" "${HRPSYS_BASE_PREFIX}/include/hrpsys/io/")
```
Comment (gergondet) [inline src/CMakeLists.txt:28]: Cleaner CMake way to do this:

```suggestion
if(NOT TARGET hrpIo::hrpIo)
  find_path(hrpIo_INCLUDE_DIR NAMES iob.h HINTS ${HRPSYS_BASE_PREFIX}/include/hrpsys/io)
  find_path(hrpIo_LIBRARY NAMES hrpIo HINTS ${HRPSYS_BASE_PREFIX})
  include(FindPackageHandleStandardArgs)
  find_package_handle_standard_args(hrpIo DEFAULT_MSG REQUIRED hrpIo_LIBRARY hrpIo_INCLUDE_DIR)
  mark_as_advanced(hrpIo_INCLUDE_DIR hrpIo_LIBRARY)
  add_library(hrpIo::hrpIo INTERFACE IMPORTED GLOBAL)
  set_target_properties(hrpIo::hrpIo PROPERTIES
    INTERFACE_INCLUDE_DIRECTORIES ${hrpIo_INCLUDE_DIR}
    INTERFACE_LINK_LIBRARIES ${hrpIo_LIBRARY}
  )
endif()
target_link_libraries(MCControl PUBLIC mc_rtc::mc_control hrpIo)
target_include_directories(MCControl PRIVATE "${PROJECT_BINARY_DIR}/idl")
```

File: src/MCControl.cpp [hunk 3]
```diff
@@ -128,6 +134,78 @@ MCControl::MCControl(RTC::Manager* manager)
...
+        coil::TimeValue coiltm(coil::gettimeofday());
+        RTC::Time tm;
+        tm.sec = static_cast<CORBA::ULong>(coiltm.sec());
+        tm.nsec = static_cast<CORBA::ULong>(coiltm.usec()) * 1000;
+        m_pgainsOut.tm = tm;
+        m_dgainsOut.tm = tm;
+        m_pgainsOutOut.write();
+        m_dgainsOutOut.write();
+#endif
+        return true;
```
Comment (gergondet) [inline src/MCControl.cpp:179]: Would be cleaner as a method that is called by the datastore rather than a long lambda

File: src/MCControl.cpp [hunk 4]
```diff
@@ -128,6 +134,78 @@ MCControl::MCControl(RTC::Manager* manager)
...
+                                                  m_pgainsInIn.read();
+                                                  m_dgainsInIn.read();
+                                                  for(unsigned int i = 0; i < m_pgainsIn.data.length(); i++)
+                                                  {
+                                                    p_vec.push_back(m_pgainsIn.data[i]);
+                                                    d_vec.push_back(m_dgainsIn.data[i]);
+                                                  }
+#endif
+                                                  return true;
+                                                });
```
Comment (gergondet) [inline src/MCControl.cpp:208]: Would be cleaner as a method that is called by the datastore rather than a long lambda

File: src/MCControl.cpp [hunk 5]
```diff
@@ -128,6 +134,78 @@ MCControl::MCControl(RTC::Manager* manager)
...
+        m_dgainsOut.tm = tm;
+        m_pgainsOutOut.write();
+        m_dgainsOutOut.write();
+#endif
+        return true;
+      });
+  controller.controller().datastore().make_call(controller.robot().name() + "::GetPDGains",
+                                                [this](std::vector<double> & p_vec, std::vector<double> & d_vec) {
+                                                  p_vec.resize(0);
+                                                  d_vec.resize(0);
```
Comment (gergondet) [inline src/MCControl.cpp:184]: wasteful in the `USE_IOB` case

File: src/MCControl.cpp [hunk 6]
```diff
@@ -128,6 +134,78 @@ MCControl::MCControl(RTC::Manager* manager)
...
+                                                  close_iob();
+#else
+                                                  // if not on real robot
+                                                  m_pgainsInIn.read();
+                                                  m_dgainsInIn.read();
+                                                  for(unsigned int i = 0; i < m_pgainsIn.data.length(); i++)
+                                                  {
+                                                    p_vec.push_back(m_pgainsIn.data[i]);
+                                                    d_vec.push_back(m_dgainsIn.data[i]);
+                                                  }
```
Comment (gergondet) [inline src/MCControl.cpp:205]: Could be resize + memcpy

File: src/MCControl.cpp [hunk 7]
```diff
@@ -128,6 +134,78 @@ MCControl::MCControl(RTC::Manager* manager)
...
     }
   }
+
+  controller.controller().datastore().make_call(
+      controller.robot().name() + "::SetPDGains",
+      [this](const std::vector<double> & p_vec, const std::vector<double> & d_vec) {
+        const auto & rjo = controller.robot().refJointOrder();
+        if(p_vec.size() != rjo.size())
+        {
+          return false;
```
Comment (gergondet) [inline src/MCControl.cpp:144]: Should print an error message too

File: src/MCControl.cpp [hunk 8]
```diff
@@ -128,6 +134,78 @@ MCControl::MCControl(RTC::Manager* manager)
...
+        }
+        if(d_vec.size() != rjo.size())
+        {
+          return false;
```
Comment (gergondet) [inline src/MCControl.cpp:148]: Should print an error message too

File: src/MCControl.cpp [hunk 9]
```diff
@@ -128,6 +134,78 @@ MCControl::MCControl(RTC::Manager* manager)
...
+        }
+
+#ifdef USE_IOB
+        // if on real robot
+        open_iob();
```
Comment (gergondet) [inline src/MCControl.cpp:153]: This can fail so you should check for it

File: src/MCControl.cpp [hunk 10]
```diff
@@ -128,6 +134,78 @@ MCControl::MCControl(RTC::Manager* manager)
...
+        for(unsigned int i = 0; i < rjo.size(); i++)
+        {
+          write_pgain(i, p_vec[i]);
+          write_dgain(i, d_vec[i]);
```
Comment (gergondet) [inline src/MCControl.cpp:157]: Those can fail as well

File: src/MCControl.cpp [hunk 11]
```diff
@@ -128,6 +134,78 @@ MCControl::MCControl(RTC::Manager* manager)
...
+        }
+        close_iob();
```
Comment (gergondet) [inline src/MCControl.cpp:159]: This can fail too

File: src/MCControl.cpp [hunk 12]
```diff
@@ -128,6 +134,78 @@ MCControl::MCControl(RTC::Manager* manager)
...
+                                                  open_iob();
+                                                  size_t num_joints = number_of_joints();
+                                                  p_vec.resize(num_joints);
+                                                  d_vec.resize(num_joints);
+                                                  for(unsigned int i = 0; i < num_joints; i++)
+                                                  {
+                                                    read_pgain(i, &p_vec[i]);
+                                                    read_dgain(i, &d_vec[i]);
+                                                  }
+                                                  close_iob();
```
Comment (gergondet) [inline src/MCControl.cpp:196]: Same remark regarding possible failure of IOB calls.

**Review by gergondet — CHANGES_REQUESTED** [review]

File: src/MCControl.cpp (hunk 3 above)
Comment (rohanpsingh) [reply src/MCControl.cpp]: Done.

**Comment by rohanpsingh** [comment]
Thanks for the review @gergondet 
I made most of the changes you requested with 2 remaining questions.
I also added functionality to read/write gains by joint name.

> I'd prefer this to be a run-time switch. We can easily detect if mc_openrtm is being run on a real robot (RobotHardware component is available) or in simulation

Is there an example for this? I'm not sure I know how to check for the `RobotHardware` component inside `mc_openrtm`

> possible failure of IOB calls.

Should I check if each call returns a `TRUE`? And what to do if `open_iob` is successful but `close_iob` is not?

**Comment by gergondet** [comment]
> Is there an example for this? I'm not sure I know how to check for the RobotHardware component inside mc_openrtm

I have implemented the member `m_is_simulation` in #25 you can rely on this to detect whether we are running in simulation or on hardware

> Should I check if each call returns a TRUE? And what to do if open_iob is successful but close_iob is not?

To make the output readable:
- if open_iob fails, a critical error message should be displayed (`mc_rtc::log::critical`) should be displayed and the feature disabled
- if setting the gains fail, same deal
- I don't think there's a scenario where `close_iob` can fail when `open_iob` didn't but it might be worth reporting if that happens

**Comment by rohanpsingh** [comment]
Thanks @gergondet !
I implemented checks for possible IOB failure calls here https://github.com/jrl-umi3218/mc_openrtm/pull/23/commits/f531cc5e8ce46ed2d01a5882ac10983b4f3044ed. It looks dirty to me, so if you have a better suggestion, please let me know.

I am also using the `m_is_simulation` member now here https://github.com/jrl-umi3218/mc_openrtm/pull/23/commits/4d0cba07a7bdd9a9628f558dce9c75020afc7f0e.

**Comment by gergondet** [comment]
Thanks @rohanpsingh 

Once the `failed_iob` issue is fixed and when this is tested on a robot we can merge.

**Comment by rohanpsingh** [comment]
Okay, I made the fix! If the CI passes, I'll find a way a to test this safely on the real robot. :smiley:

**Comment by rohanpsingh** [comment]
I confirmed that I could change the gains on the real robot (with servo ON)

**Pull request merged**